fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveRemoteFileIO.java

```diff
@@ -107,16 +108,72 @@ public Map<RemotePathKey, List<RemoteFileDesc>> getRemoteFiles(RemotePathKey pat
         return resultPartitions.put(pathKey, fileDescs).build();
     }
 
+    private RemoteIterator<LocatedFileStatus> listFilesRecursive(FileSystem fileSystem, Path f)
+        throws FileNotFoundException, IOException {
+        return new RemoteIterator<LocatedFileStatus>() {
+            private Stack<RemoteIterator<LocatedFileStatus>> itors = new Stack<>();
+            private RemoteIterator<LocatedFileStatus> curItor = fileSystem.listLocatedStatus(f);
+            private LocatedFileStatus curFile;
+
+            @Override
+            public boolean hasNext() throws IOException {
+                while (curFile == null) {
+                    if (curItor.hasNext()) {
+                        handleFileStat(curItor.next());
+                    } else if (!itors.empty()) {
+                        curItor = itors.pop();
+                    } else {
+                        return false;
+                    }
+                }
+                return true;
+            }
+
+            // Process the input stat.
+            // If it is a file, return the file stat.
+            // If it is a valid directory, traverse it.
+            private void handleFileStat(LocatedFileStatus stat) throws IOException {
+                if (stat.isFile()) {
+                    curFile = stat;
+                } else if (isValidDirectory(stat)) {
+                    try {
+                        RemoteIterator<LocatedFileStatus> newDirItor = fileSystem.listLocatedStatus(stat.getPath());
+                        itors.push(curItor);
+                        curItor = newDirItor;
+                    } catch (FileNotFoundException ignored) {
+                        LOG.debug("Directory {} deleted while attempting for recursive listing", stat.getPath());
+                    }
+                }
+            }
+
+            @Override
+            public LocatedFileStatus next() throws IOException {
+                if (hasNext()) {
```

Comment:
hasNext() run twice

Reply:
> hasNext() run twice

The first hasNext set the curFile and the second one only check it and usually return true directly. So it's ok.